### PR TITLE
refac: Add interface sub-project to master data module

### PR DESCRIPTION
**Pull request by MWO1024**
## Description

Introduce an interfaces and application sub projects to the master data module, similar to the other modules. It contains a client to depend on for master data related functionality.
Additionally, all master data related functionality has been moved to the master data module. The change should have no functional inpact.

## References

This is a continuation of #670.

### CODE REVIEW

**Comment by MWO1024** [comment]
Aside from all the usual, please consider the following as well:
* Should we use the private assets for our modules or not?
* Is the visibility sensible? Some things have been marked as internal, and more probably can.
* When should we commit the master data context? I created a manager that handles the commits and is invoked by the client. Perhaps a better pattern exists?
* The name of some of the dtos is a bit weird. The name from the command was used verbatim, but its hjighly likely a better name can be found for some of them.

File: source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/IIntegrationEventHandler.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+// limitations under the License.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Energinet.DataHub.Core.Messaging.Communication;
+
+namespace Energinet.DataHub.EDI.Infrastructure.Configuration.IntegrationEvents.IntegrationEventMappers;
+
+/// <summary>
+///     SDometgasdgastgas
```
Comment (lasrinnil) [inline source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/IIntegrationEventHandler.cs:22]: 😃

File: source/IntegrationTests/Application/Actors/CreateActorsTests.cs [hunk 2]
```diff
@@ -51,9 +53,29 @@ public async Task Actor_is_created()
...
-    private static CreateActorCommand CreateCommand()
+    [Fact]
+    public async Task Actor_is_not_created_multiple_times()
+    {
+        var dto1 = CreateDto();
+        var dto2 = CreateDto();
+        var dto3 = CreateDto();
+        var dto4 = CreateDto();
+
+        await _masterDataClient.CreateActorIfNotExistAsync(dto1, CancellationToken.None);
```
Comment (lasrinnil) [inline source/IntegrationTests/Application/Actors/CreateActorsTests.cs:64]: Is there a reason to call this more than twice?

File: source/IntegrationTests/Application/Actors/CreateActorsTests.cs [hunk 3]
```diff
@@ -63,6 +85,18 @@ private async Task<Actor> GetActor()
         return await connection.QuerySingleOrDefaultAsync<Actor>(sql);
     }
 
+    private async Task<IEnumerable<Actor>> GetAllActorsWithId()
```
Comment (lasrinnil) [inline source/IntegrationTests/Application/Actors/CreateActorsTests.cs:88]:
```suggestion
    private async Task<IEnumerable<Actor>> GetAllActors()
```

File: source/MasterData.Application/Configuration/MasterDataConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+public static class MasterDataConfiguration
+{
+    public static void AddMasterDataModule(this IServiceCollection services, IConfiguration configuration)
+    {
+        services.AddScopedSqlDbContext<MasterDataContext>(configuration);
+
+        services.AddScoped<MasterDataTransactionManager, MasterDataTransactionManager>();
+
+        services.AddScoped<IGridAreaRepository, GridAreaRepository>();
```
Comment (lasrinnil) [inline source/MasterData.Application/Configuration/MasterDataConfiguration.cs:39]: why not transiant?

File: source/MasterData.Application/MasterData.Application.csproj [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <AssemblyName>Energinet.DataHub.EDI.MasterData.Application</AssemblyName>
+    <RootNamespace>Energinet.DataHub.EDI.MasterData.Application</RootNamespace>
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Application\Application.csproj" />
```
Comment (lasrinnil) [inline source/MasterData.Application/MasterData.Application.csproj:7]: What is used from `Application` still? Just curious 😄

File: source/MasterData.Interfaces/IMasterDataClient.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// <summary>
+    ///     Gets the <see cref="ActorNumberAndRoleDto" /> associated with the given certificate thumbprint, if any.
+    /// </summary>
+    Task<ActorNumberAndRoleDto?> GetActorNumberAndRoleFromThumbprintAsync(CertificateThumbprint thumbprint);
+
+    /// <summary>
+    ///     Get the <see cref="ActorNumber"/> of DataHub.
+    /// </summary>
+    static ActorNumber DataHubIdentificationNumber()
```
Comment (lasrinnil) [inline source/MasterData.Interfaces/IMasterDataClient.cs:64]: This seems as Masterdata domain information - what do you think?

File: source/MasterData.Interfaces/Models/GridAreaOwnershipAssignedDto.cs [hunk 7]
```diff
@@ -13,26 +13,12 @@
...
-    public Instant ValidFrom { get; }
-
-    public ActorNumber GridAreaOwnerActorNumber { get; }
-
-    public int SequenceNumber { get; }
-}
+public sealed record GridAreaOwnershipAssignedDto(
+    string GridAreaCode,
+    Instant ValidFrom,
+    ActorNumber GridAreaOwnerActorNumber,
```
Comment (lasrinnil) [inline source/MasterData.Interfaces/Models/GridAreaOwnershipAssignedDto.cs:23]:
```suggestion
    ActorNumber GridAreaOwner,
```

File: source/MaterData.Domain/MaterData.Domain.csproj [hunk 8]
```diff
@@ -5,5 +5,6 @@
   </PropertyGroup>
   <ItemGroup>
     <ProjectReference Include="..\BuildingBlocks.Domain\BuildingBlocks.Domain.csproj"/>
+    <ProjectReference Include="..\MasterData.Interfaces\MasterData.Interfaces.csproj"/>
```
Comment (lasrinnil) [inline source/MaterData.Domain/MaterData.Domain.csproj:8]: This reference should not be here

File: source/Tests/CimMessageAdapter/Messages/RequestAggregatedMeasureData/MessageParserTests.cs [hunk 9]
```diff
@@ -234,7 +234,7 @@ private static MemoryStream CreateProtoMemoryStream()
         var request = new RequestAggregatedMeasureDataDto(
             "5799999933318",
             "DDK",
-            DataHubDetails.IdentificationNumber.Value,
+            IMasterDataClient.DataHubIdentificationNumber().Value,
```
Comment (lasrinnil) [inline source/Tests/CimMessageAdapter/Messages/RequestAggregatedMeasureData/MessageParserTests.cs:237]: Why is it named "IdentificationNumber"? Isn't it "just" a `ActorNumber`

**Review by lasrinnil — COMMENTED** [review]
Great job 👍

File: source/MasterData.Application/MasterData.Application.csproj (hunk 5 above)
Comment (MWO1024) [reply source/MasterData.Application/MasterData.Application.csproj:7]: Nothing it appears. The reference is going to be removed with the upcoming review changes.

File: source/MasterData.Interfaces/IMasterDataClient.cs (hunk 6 above)
Comment (MWO1024) [reply source/MasterData.Interfaces/IMasterDataClient.cs:64]: That was the initial thought, but it is used outside of the master data module (its only used outside of the master data module iirc). Not sure how to alliviate that in a sensible way.

File: source/Api/IntegrationEventHandler.cs [hunk 10]
```diff
@@ -45,21 +46,29 @@ public async Task HandleAsync(IntegrationEvent integrationEvent)
     {
         ArgumentNullException.ThrowIfNull(integrationEvent);
 
-        var shouldHandleEvent = _integrationEventMappers.TryGetValue(integrationEvent.EventName, out var integrationEventMapper);
+        var canHandleEvent =
+            _integrationEventMappers.TryGetValue(integrationEvent.EventName, out var integrationEventMapper);
 
-        if (!shouldHandleEvent)
+        if (!canHandleEvent || integrationEventMapper is null)
```
Comment (lasrinnil) [inline source/Api/IntegrationEventHandler.cs:52]: could we simplify this further by simply opting out here if integrationEventMapper is null?

File: source/IntegrationTests/Application/Actors/CreateActorsTests.cs [hunk 11]
```diff
@@ -13,36 +13,38 @@
...
-        _mediator = GetService<IMediator>();
+        _masterDataClient = GetService<IMasterDataClient>();
         _connectionFactory = GetService<IDatabaseConnectionFactory>();
     }
 
     [Fact]
     public async Task Actor_is_created()
     {
-        var command = CreateCommand();
+        var dto = CreateDto();
```
Comment (lasrinnil) [inline source/IntegrationTests/Application/Actors/CreateActorsTests.cs:45]: please use the variable name to identify the type created from the `CreateDto()` method

File: source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/IIntegrationEventMapper.cs [hunk 12]
```diff
@@ -12,10 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+using System.Threading;
 using System.Threading.Tasks;
 using Energinet.DataHub.Core.Messaging.Communication;
-using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
-using MediatR;
 
 namespace Energinet.DataHub.EDI.Infrastructure.Configuration.IntegrationEvents.IntegrationEventMappers;
```
Comment (lasrinnil) [inline source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/IIntegrationEventMapper.cs:19]: If this no longer is a mapper - perhaps the namespace should be updated to reflect that as well

File: source/IntegrationTests/Infrastructure/Configuration/IntegrationEvents/TestIntegrationEventMapper.cs [hunk 13]
```diff
@@ -23,14 +24,23 @@ namespace Energinet.DataHub.EDI.IntegrationTests.Infrastructure.Configuration.In
...
 
-    public Task<ICommand<Unit>> MapToCommandAsync(IntegrationEvent integrationEvent)
+    public async Task HandleAsync(IntegrationEvent integrationEvent, CancellationToken cancellationToken)
     {
         MappedCount++;
 
-        return Task.FromResult<ICommand<Unit>>(new TestCommand());
+        var fromResult = await Task.FromResult<ICommand<Unit>>(new TestCommand());
+
+        await _mediator.Send(fromResult, cancellationToken).ConfigureAwait(false);
```
Comment (lasrinnil) [inline source/IntegrationTests/Infrastructure/Configuration/IntegrationEvents/TestIntegrationEventMapper.cs:44]: is mediator still being used with integrationevents

File: source/MasterData.Application/Configuration/MasterDataConfiguration.cs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+public static class MasterDataConfiguration
+{
+    public static void AddMasterDataModule(this IServiceCollection services, IConfiguration configuration)
+    {
+        services.AddScopedSqlDbContext<MasterDataContext>(configuration);
+
+        services.AddTransient<IGridAreaRepository, GridAreaRepository>();
+        services.AddTransient<IDataRetention, GridAreaOwnerRetention>();
+
```
Comment (lasrinnil) [inline source/MasterData.Application/Configuration/MasterDataConfiguration.cs:38]: nit: remove the whitespaces

File: source/IntegrationTests/Infrastructure/Configuration/IntegrationEvents/TestIntegrationEventMapper.cs (hunk 13 above)
Comment (MWO1024) [reply source/IntegrationTests/Infrastructure/Configuration/IntegrationEvents/TestIntegrationEventMapper.cs:44]: For CalculationResultCompleted, yes. But that is the last one, and one could make the argument, that when swap to the new EnergyResultProducedV2, we fix this in the process.

File: source/MasterData.Application/MasterData.Application.csproj [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\BuildingBlocks.Application\BuildingBlocks.Application.csproj"/>
+    <ProjectReference Include="..\BuildingBlocks.Infrastructure\BuildingBlocks.Infrastructure.csproj"/>
+    <ProjectReference Include="..\MasterData.Interfaces\MasterData.Interfaces.csproj"/>
+    <ProjectReference Include="..\MaterData.Infrastructure\MaterData.Infrastructure.csproj"/>
+  </ItemGroup>
+  <ItemDefinitionGroup>
+    <ProjectReference>
+      <PrivateAssets>all</PrivateAssets>
```
Comment (lasrinnil) [inline source/MasterData.Application/MasterData.Application.csproj:14]: Is this relevant to us?

File: source/MasterData.Interfaces/MasterData.Interfaces.csproj [hunk 16]
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <AssemblyName>Energinet.DataHub.EDI.MasterData.Interfaces</AssemblyName>
+    <RootNamespace>Energinet.DataHub.EDI.MasterData.Interfaces</RootNamespace>
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\BuildingBlocks.Domain\BuildingBlocks.Domain.csproj"/>
+    <ProjectReference Include="..\MaterData.Domain\MaterData.Domain.csproj"/>
```
Comment (lasrinnil) [inline source/MasterData.Interfaces/MasterData.Interfaces.csproj:8]: This reference should not be here

File: source/MasterData.Application/MasterDataClient.cs [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+        return _actorCertificateRepository
+            .AddAsync(newCertificate, cancellationToken);
+    }
+
+    private void UpdateExistingActorCertificate(
+        ActorCertificateCredentialsAssignedDto request,
+        ActorCertificate existingCertificate)
+    {
+        var updateError = existingCertificate.Update(
```
Comment (lasrinnil) [inline source/MasterData.Application/MasterDataClient.cs:147]: ☎️

File: source/IntegrationTests/Infrastructure/Configuration/IntegrationEvents/TestIntegrationEventMapper.cs (hunk 13 above)
Comment (MWO1024) [reply source/IntegrationTests/Infrastructure/Configuration/IntegrationEvents/TestIntegrationEventMapper.cs:44]: I have added a scribble on my whiteboard about this.

**Review by lasrinnil — APPROVED** [review]

File: source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/IIntegrationEventProcessor.cs [hunk 18]
```diff
@@ -32,5 +31,5 @@ public interface IIntegrationEventMapper
     /// <summary>
     /// Process a single integration event
     /// </summary>
-    public Task<ICommand<Unit>> MapToCommandAsync(IntegrationEvent integrationEvent);
+    public Task HandleAsync(IntegrationEvent integrationEvent, CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (MWO1024) [reply source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/IIntegrationEventProcessor.cs]: Changed to processAsync, to match the interface name.

**Review by lasrinnil — APPROVED** [review]
🎆 G O O D 🎆  S T U F F 🎆
